### PR TITLE
Improve type inference for $_

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes the type inference for $_ in pipelines with more than 1 command by making it infer the type of the previous command of the expression rather than the first.

-Edit: Decided to make a full fix for $_ and also fix it when the scriptblock is inside an array of arguments like for Select-Object or Foreach-Object `ls | ForEach-Object -Process {$_.<Tab>},{$_.<Tab>}`  
Also fixes a scenario where a catch block is given invalid types so there's no completion for $_ inside the catch block.
Fixes: https://github.com/PowerShell/PowerShell/issues/17613
Edit2: The test failures helped me discover a different issue, strings are enumerated as chars by the type inference. This has now been fixed.

Full list of scenarios where $_ is now properly being inferred:

- Member invocation scriptblocks like `$Collection.Where({$_})`
- Scriptblocks used as command arguments like: `ls | where {$_}`
- Switch statements like: `switch ("hello") {'Hello'{$_}}`
- Catch statements like: `try{} catch {$_}`
- Trap statements like: `trap {$_}`

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Review by iSazonov — APPROVED** [review]

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 1]
```diff
@@ -1840,82 +1840,68 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
...
                     {
-                        // Script block in a hash table, could be something like:
-                        //     dir | ft @{ Expression = { $_ } }
-
-                        if (parent.Parent.Parent.Parent is HashtableAst)
+                        hasSeenScriptBlock = true;
+                    }
+                    else if (hasSeenScriptBlock)
+                    {
+                        if (parent is InvokeMemberExpressionAst invokeMember)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:1856]: Can you please add a comment here explaining what this scenario is (with an example)?

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 2]
```diff
@@ -1840,82 +1840,68 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
...
                         {
-                            parent = parent.Parent.Parent.Parent;
+                            parent = invokeMember.Expression;
+                            break;
                         }
-                        else if (parent.Parent.Parent.Parent is ArrayLiteralAst && parent.Parent.Parent.Parent.Parent is HashtableAst)
+                        else if (parent is CommandAst cmdAst && cmdAst.Parent is PipelineAst pipeline && pipeline.PipelineElements.Count > 1)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:1861]: Can you please also add a comment here?

**Review by daxian-dbw — COMMENTED** [review]
Looks good, but it would be great if more comments can be added.

**Comment by MartinGC94** [comment]
Sorry for pushing new updates after the review, I forgot that there were other instances where `$_` is set. Hopefully I have them all now.

**Review by daxian-dbw — APPROVED** [review]
Looks great! Thank you, @MartinGC94!

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 3]
```diff
@@ -1840,82 +1840,101 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
...
+
+                if (parent is CatchClauseAst catchBlock)
+                {
+                    if (catchBlock.CatchTypes.Count > 0)
+                    {
+                        foreach (TypeConstraintAst catchType in catchBlock.CatchTypes)
                         {
-                            inferredTypes.Add(new PSTypeName(typeof(ErrorRecord)));
+                            Type exceptionType = catchType.TypeName.GetReflectionType();
+                            if (exceptionType != null && typeof(Exception).IsAssignableFrom(exceptionType))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:1904]:
```suggestion
                            if (typeof(Exception).IsAssignableFrom(exceptionType))
```

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 4]
```diff
@@ -1840,82 +1840,101 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
...
+                else if (parent is TrapStatementAst trap)
+                {
+                    if (trap.TrapType is not null)
                     {
-                        // We found a command, see if there is a previous command in the pipeline.
-                        PipelineAst pipelineAst = (PipelineAst)commandAst.Parent;
-                        var previousCommandIndex = pipelineAst.PipelineElements.IndexOf(commandAst) - 1;
-                        if (previousCommandIndex < 0)
+                        Type exceptionType = trap.TrapType.TypeName.GetReflectionType();
+                        if (exceptionType is not null && typeof(Exception).IsAssignableFrom(exceptionType))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:1922]:
```suggestion
                        if (typeof(Exception).IsAssignableFrom(exceptionType))
```

**Review by iSazonov — APPROVED** [review]

**Pull request merged**